### ISSUE TITLE
Custom Image Size Only Works in Editor, Not Public Page

### ISSUE THREAD

**Issue by mountaineer20** [pojome/elementor#4679]
I'm a paid Elementor user for multiple sites, but this error is in regards to the basic Elementor image element. The one used to display photos in posts. I have been using the custom image size option to ensure that images are the same aspect ratio and look good when positioned next to each other.

This method works great when I am composing a post in the Elementor editor. Works exactly how I’d expect it to. The problem comes up when I switch over to the public view, what anyone visiting my webpage will see. The public page reverts back to the original, non-cropped image, which looks ugly.

Examples of good vs bad are shown here: https://photos.app.goo.gl/008ZFxWTZYcqHx6Z2

An example of the issue on my page is [here](http://michaelingmanson.com/falcon-heavy-launch/).

This issue is also described [here](https://github.com/pojome/elementor/issues/4390) and [here](https://wordpress.org/support/topic/custom-image-size-only-works-in-editor-not-public-page/#post-10362036).

Thanks,
-Michael

**Comment by bainternet** [comment]
Please contact our support at support@elementor.com for guidance in resolving this issue.

**Comment by mountaineer20** [comment]
OK, thanks. I'll reach out to them for instructions and guidance on fixing this.

**Comment by pingram3541** [comment]
@bainternet Hi Raz, this is a duplicate of https://github.com/pojome/elementor/issues/4390 that is yet to be fixed, fyi.

**Comment by mountaineer20** [comment]
@pingram3541 I got things working for me at least. I did the following:

1)	Updated all plugins & themes
2)	Disabled all plugins except Elementor
3)	Switched my theme to the 2017 Wordpress default
4)	Switched my theme back to OceanWP
5)	Re-enabled plugins

Not everything works as expected for me.

**Comment by pingram3541** [comment]
@mountaineer20 Thanks for the suggestions but in the case with ALL of my sites historically this has never worked and I'm not entirely convinced it's fixed for you either.  Did you actually inspect the image loaded on the front end and verify it was NOT the original file and not just resized with css rules or the container itself?

For example, here is a fresh installed WordPress site.  You can see I am choosing a custom size and in the preview, it is indeed a completely different file and url of that file being used.

![screen shot 2018-06-20 at 12 25 01 pm](https://user-images.githubusercontent.com/319284/41680244-ba7f8656-7485-11e8-9c3a-60ed91d77c99.png)

The images on the front end when checked in the inspector are indeed the full images and NOT the resized images (see image url and dimension info):

![screen shot 2018-06-20 at 12 26 11 pm](https://user-images.githubusercontent.com/319284/41680282-d26d5766-7485-11e8-8c78-bb47cf6f464c.png)

vs what is seen in the Elementor Editor:

![screen shot 2018-06-20 at 12 25 37 pm](https://user-images.githubusercontent.com/319284/41680252-c1c7ebe2-7485-11e8-916f-167693403d59.png)

I've tested this in many hosting environments and always the same result and never has worked for me...ever.

This results in the same kb/mb file size as the original directly affecting page load.

I have since chosen not to rely on this feature and duplicate my images and upload the sizes I need.  If this feature were fixed it would save me countless hours of time.

1. Updated all plugins & themes - yep `E-v2.0.16`, `Epro-v2.0.11` as of 6/20/18
2. Disabled all plugins except Elementor - only Elementor and Elementor Pro active (ever)
3. Switched my theme to the 2017 Wordpress default - using Elementor's `Hello` theme unmodified (tried 2017 as well)
4. Switched my theme back to OceanWP - switched between Hello and 2017 and back
5. Re-enabled plugins - disabled both Elementor plugins and reenabled
6. Verified no cache using new browser at a new machine
7. Again no other plugins except Elementor so no caching via WP, no caching at the host and always cleared cache with hard reload in the browser...beside even with cache it will eventually update, this never changes.

**Comment by pingram3541** [comment]
opened a Pro support ticket...will report back.

**Comment by mountaineer20** [comment]
I spoke too soon - It's not broken again on my site - http://michaelingmanson.com/road-trip-to-nowhere/

:(

**Comment by pingram3541** [comment]
Pro support responded and noted that my home url and site url in the system info file were not the same.  I checked both home url and site url in WordPress Settings and they were both exactly the same, like:

```
WordPress Address (URL): http://mysite.com
Site Address (URL): http://mysite.com
```

I then looked at the system info file and one was https while the other http, like:

```
Site URL:	https://mysite.com	
Home URL: http://mysite.com
```

This is due to how Cloudflare's Flex SSL works.  All urls on the front end are delivered to the browser over https but the website must run under http as no certificate exists on the local web server.

The problem is not http vs https directly, even though that likely plays a role in how a filter is configured maybe (regex for ex), one can clearly see the file paths and image names are completely different.  That is at the core of the problem.

However I did test this on an http only site without Cloudflare active and I did get it to work on the front end so we may have found the bug.  Hopefully it's a quick fix to check for https or Cloudflare plugin to render the correct url.   I have tons of my clients that use Cloudflare free SSL and asking them all to go purchase Full SSL certs is not something I want to head up at the moment.

I responded to the ticket about 17hrs ago, hopefully they find a way to work around this...may even dig into it a bit myself if I get some free time.  I'll report back how it goes.

**Comment by drawcard** [comment]
Thanks for staying on top of this @pingram3541 I would be keen to hear what your findings are :)

**Comment by pingram3541** [comment]
So I followed the rabbit down the hole a bit and found...the `render` function and then... `get_settings_for_display`...then to...`Group_Control_Image_Size` class.. and ultimately would up at an added library found in `/elementor/includes/libraries/bfi-thumb/bfi-thumb.php`

The issue comes down to whether the url returns "truthy" or not starting on line 443:

When the front end is executed, and the site is using Cloudflare's Flex SSL, the image `$url` var contains `https` and does not match the `site url` which is `http` (somewhere further down the rabbit's hole) and thus does not meet either of the "if" conditions that follow, which when truthy returns the "resized" image url path.

When the back end is executed, the `$url` var does not contain `https` because Cloudflare does not rewrite ajax requests (even though the editor is indeed served over https) and thus returns truthy and passes one of the "if" conditions resulting in a properly resized image url, but of course without being rewritten by Cloudflare having `https`, which is seen in my screen shots above, notice the resized back end image url is `http`.

If I add to line 442 for example, a simple snippet that rewrites https to http:

`$url = str_replace('https', 'http', $url);`

This changes the url to truthy and...

Both the front end and back end result in the expected resized image.  My initial thoughts would be to somehow place a check to see if Cloudflare Flex SSL is being used (since it requires the Cloudflare plugin to also be installed and active to work properly), we can mitigate this problem.

For now a hard hack would be to add that str_replace above line 443 but this is a core hack atm.

**Comment by mountaineer20** [comment]
Have you heard anything back from the pro support ticket? When might this be fixed and rolled out?

**Comment by pingram3541** [comment]
No response at all, it's been 3 days since I responded to their first and only response to my ticket which was very dismissive imo.  I responded about 3hrs after they emailed me.  No other communication since.

I logged in yesterday before I put the work in above, only to find they used the credentials I provided them to log in and disable the Cloudflare plugin.  They did not tell me they did this nor did they ask if it was ok first. (it's a dev site so I really don't care but it broke the Flex SSL and caused external images, scripts, etc to load insecurely -- but it did fix the front end image paths which is what led me to do the work above) ...but what's the point, if not telling me what they did?

I've never been impressed with Pro support and honestly try my best to never use it, each new experience has resurfaced the same emotions from the previous unfortunately.  I don't feel like I'm working with developers that know the code intimately but rather more like gatekeepers.

Hopefully enough feedback will promote change and we'll get some quality support.  For now I buy the Pro license for everything **BUT** the support and I honestly get frustrated when forced to use Pro support on the wp forums, especially when it has nothing to do with Pro...they see it is installed and instantly that is the route one must take, [especially if the wp.org mods see it](https://wordpress.org/support/topic/feedback-about-closing-topics-presumptuously/).

Maybe @benpines can see this **NOT** as a dig against the team but as critical feedback from a concerned perspective that we need better support/better communication or we won't use it and even worse, desperately try to avoid it.  Anyway, it's all good big picture-wise and I have no ill feelings towards anyone on the team and I feel now more than ever one must tread lightly when it comes to feelings but I just want to have a solid and reliable product for my clients...which it already does a very good job of.  But given the opportunity, I'd much rather be in the position of singing praises about their Pro support and I hope very soon they'll give me the opportunity to do so.

**Comment by drawcard** [comment]
Appreciate the level of research you're doing @pingram3541 - I hope the dev team can get to the bottom of it since it's going to affect numerous users using CloudFlare (my host has CF integrated into the shared hosting accounts used by my clients, and there's no ability to switch it off).

**Comment by pingram3541** [comment]
@drawcard Did you try the snippet I posted above to see if it works for you?

I copied the original `bfi-thumb.php` and archived it and I also keep a zipped version of the edited file just in case an update overwrites it.  I know it's a core edit but hopefully only temporary and just might save you a ton of unnecessary work.

For the record, it doesn't look like the [bfi_thumb repo](https://github.com/bfintal/bfi_thumb) is being maintained anymore.

**Comment by bainternet** [comment]
A Few corrections:

> This is due to how Cloudflare's Flex SSL works. All urls on the front end are delivered to the browser over https but the **website must run under http** as no certificate exists on the local web server.

You can server both front and back ends over https using CloudFlare Flexiable SSL which is something I personaly have running on a few site. The whole idea of Flexiable SSL is to be used when no certificate exists on the origin web server regardless of which software is running.

Second this issue is not an elementor issue. Elementor uses BFI_Thumb which in turn uses the native WP functions to generate cropped version and gets the url from native WP functions as well.

So do define this correctly, this is an issue with specific envoirment setups where WP site/home urls are set as HTTP and only cloudflare is defined to server over HTTPS.

### Possible solutions:

- The correct solution whould be to have a real SSL certificate defined on the origin server and performing a real migration to HTTPS for the entire site. Which I understand that sadly is not always a valid options when it comes to clients.

- Keep using CloudFlare Flexiable but performe a real migration to HTTPS.

- Like BFI_Thumbs works, use the provided native WP hooks to make modifications to urls as needed, something like:

```PHP
/**
 * @param array|false  $image         Either array with src, width & height, icon src, or false.
 * @param int          $attachment_id Image attachment ID.
 * @param string|array $size          Size of image. Image size or array of width and height values
 *                                    (in that order). Default 'thumbnail'.
 * @param bool         $icon          Whether the image should be treated as an icon. Default false.
 */
add_filter( 'wp_get_attachment_image_src', function ( $image, $attachment_id, $size, $icon ) {
	// Check that image is not false
	if ( ! $image ) {
		return $image;
	}
	// Check that this is a bfi_thumb related image
	if ( empty( $size['bfi_thumb'] ) ) {
		return $image;
	}
	// Check that CloudFlare is serving via https
	if ( empty( $_SERVER['HTTP_X_FORWARDED_PROTO'] ) || 'https' !== $_SERVER['HTTP_X_FORWARDED_PROTO'] ) {
		return $image;
	}

	// Check if the site is set to a different protocol
	if ( false !== strpos( site_url(), 'https' ) ) {
		return $image;
	}

	if ( ! empty( $image['src'] ) ) {
		// force http protocol
		list( $protocol, $uri ) = explode( '://', $image['src'], 2 );
		$image['src'] = 'http://' . $uri;
	}

	return $image;
} );
```

Bare in mind that this is not tested but should work.

and as for "performe a real migration to HTTPS" I mean full search-replace for the eintire site.

let me know.

**Comment by pingram3541** [comment]
Thanks for the response Raz.  A few thoughts if I may...

> You can server both front and back ends over https using CloudFlare Flexible SSL which is something I personally have running on a few sites. The whole idea of Flexible SSL is to be used when no certificate exists on the origin web server regardless of which software is running.

You are suggesting to change all my urls to `https` via find & replace but this will break my site with the Flex SSL service.  Flex SSL cannot have `https` between Cloudflare and the origin server as you say.  It will result in redirect loops.

https://support.cloudflare.com/hc/en-us/articles/115000219871-Why-does-Flexible-SSL-cause-a-redirect-loop-

> Second, this issue is not an elementor issue. Elementor uses BFI_Thumb which in turn uses the native WP functions to generate cropped version and gets the URL from native WP functions as well.

BFI_Thumb is a 3rd party library Elementor uses not WP core which is why I opened the bug report here.  The issue is experienced in Elementor for anyone using Cloudflare's Flex SSL the way Cloudflare suggests the WP environment to be setup for the Flex SSL product.

> So to define this correctly, this is an issue with specific environment setups where WP site/home URLs are set as HTTP and only Cloudflare is defined to serve over HTTPS.

Yes, this is how Cloudflare's Flex SSL works.

> The correct solution would be to have a real SSL certificate defined on the origin server and performing a real migration to HTTPS for the entire site. Which I understand that sadly is not always a valid option when it comes to clients.

Why is the CORRECT solution?  Are you suggesting that Cloudflare FlexSSL is not a viable product for use with Elementor when the problem can be fixed with a simple filter to BFI_Thumb?

> Keep using CloudFlare Flexible but perform a real migration to HTTPS.

This is impossible in regards to Flex SSL and essentially identical to the first suggestion above, to overhaul the server and NOT use Flex SSL but rather Full SSL.

> Like BFI_Thumbs works, use the provided native WP hooks to make modifications to urls as needed, something like...

This is awesome stuff, thanks for taking the time to provide this example.

I'm just not sure it is necessary for every wp instance when we only need to do this when BFI_Thumb calls this function and I'll test to see if it indeed works but are you then suggesting that everyone using Elementor and Cloudflare's Flex SSL add this to their functions.php?  This doesn't seem like a user friendly fix for such a popular product being used on the internet today.  Wouldn't it be better to tailor BFI_Thumb for what the internet is now doing as a fairly popular thing?

Thanks again for looking into this.  We just now need to determine if we can be proactive and provide a simple fix for everyone using Elementor and Cloudflare's Flex SSL or make the bold statement that Elementor cannot no longer support users with Cloudflare's Flex SSL and they must use Full SSL or No SSL on their servers for the image resize function to work.

Thx for hearing my thoughts.  Cheers!

**Comment by bainternet** [comment]
@pingram3541 

> Flex SSL **cannot** have https between Cloudflare and the origin server as you say. It will result in redirect loops.

Again, Yes you can, that is why ColudFlare's own plugin implements a "Fix" for that "redirect loop". 

Now let's agree that having the site configured as HTTP and serving it over HTTPS is less than ideal and if you are looking for a "bold" statement then: 

**Elementor "custom sized images" will not work correctly out of the box using Cloudflare's Flexible SSL to serve the site over HTTPS when the site URL in WordPress is set to HTTP.**

> provide a simple fix for **everyone** using Elementor and Cloudflare's Flex SSL

If you mean anyone that is using Cloudflare's Flexible SSL to serve the site over HTTPS when the site URL in WordPress is set to HTTP, which is not a valid setup, then we must to think about **everyone** else, and by that I mean, that if we implement the filter we will add an unnecessary filter to all other sites that don't use Cloudflare's Flexible SSL or even those that do use it but have the site configured as HTTPS as well.

Now you keep saying that a site owner cannot set up the site to https when he uses Cloudflare's Flexible SSL, well this is a wrong statement.
Yes, if you do not have Cloudflare's plugin or one of the many 3rd party plugins out there that implement a simple "fix", then you will have a redirect loop, but that does not mean that it can't work.

#### Side note:
The "redirection loop fix" is to tell cloudflare that the site is responding in https ex:

```PHP
// for cloudflare
if ( isset( $_SERVER[ 'HTTP_CF_VISITOR' ] ) && ( strpos( $_SERVER[ 'HTTP_CF_VISITOR' ], 'https' ) !== false ) ) {
    $_SERVER[ 'HTTPS' ] = 'on';
}
 
// for other...
if(isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && $_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https'){
    $_SERVER['HTTPS']='on';
}
```

As I stated before, I personally use Cloudflare's Flexible SSL with the site URL set as HTTPS on many sites and when I set it up I instantly add the above lines to my wp-config.php, Now I'm not asking you to do that or anyone for that manner, but if your site is configured differently from the way it is served then yes, you would need to have a custom filter hook to get Elementor's custom sized images working correctly.

So the main question here is, Should this hook be implemented in Elementor or not? And the answer is, from my point of view, that the vast majority of our users don't need this implemented. I could be wrong (it happens from time to time), but how can we know for sure?

Thanks for contributing in this civilized discussion :)

**Comment by pingram3541** [comment]
Hey @bainternet sorry for the long delay I was waiting on a ticket response from Cloudflare and it took a couple back and forth threads to get the confirmation I was looking for.

You are 100% correct and they no longer require the site url in Wordpress settings to be http when the local server is without a certificate installed and using the Flexible SSL service.

They also confirmed I was 100% correct in that this did not used to be the case but that the Cloudflare plugin has been updated to fix this problem since.  They also confirmed that their documentation doesn't really clarify this because now it doesn't matter, from their perspective, when using the Cloudflare plugin and Flexible SSL, whether the Wordpress site url is http or https, the plugin makes adjustments as needed either way...except in the case here with BFI_thumb.

I changed my site urls to https and all is good.  No redirect loops and proper resized images are in effect.  Thank you so much for your patience with me and I'm glad we were able to flush this all out and hopefully save others some time.

**Comment by bainternet** [comment]
@pingram3541 
So glad we figured this out.

@mountaineer20 
to quote @pingram3541 

> I changed my site URLs to https and all is good.  No redirect loops and proper resized images are in effect.

**Comment by mountaineer20** [comment]
@bainternet Is there a menu option to push the HTTPS chance sitewide? When I visit https://michaelingmanson.com/falcon-heavy-launch/ I still experience the issue.

**Comment by pingram3541** [comment]
@mountaineer20 From fresh install perspective:

Wordpress:
1) Wordpress Address url and Site Address url as `http` initially (Settings > General)
2) Install and activate the Cloudflare plugin
3) Register/Authenticate Cloudflare plugin and choose Optimize Cloudflare for Wordpress

Cloudflare:
4) On DNS tab, make sure your site's domain A/CNAME record has an orange cloud (enabled)
5) On Crypto tab, set the SSL as type Flexible
6) On Crypto tab, enable Always use HTTPS and Automatic HTTPS Rewrites
7) On Crypto tab, once a Universal Shared certificate is showing in the Edge Certificates area...
* may take up to 24hrs to generate for new account

back on Wordpress:
8) Change the Wordpress Address and Site Address urls to https

If you see errors in browser console or red https indicator in browser, it is likely because the CF account is too new and an edge cert is not yet ready

Also, please keep in mind that if you disable the Cloudflare plugin, your site may become inaccessible without having to access your database and manually change these fields back to http.  To avoid this, always change your Wordpress urls back to http before disabling the Cloudflare plugin.

This is another reason why I tried to avoid setting it up this way as my customers will never know this.  Oh well, least friction for now and it's working.  Cheers!

**Comment by Lukasusanj** [comment]
Quick fix for this issue within **Wordpress** is that if you're using **Jetpack plugin,** **don't enable Image Performance feature within the plugin.**

This feature serves images from the server location nearest to your visitors.

They are using a global content delivery network, which can ignore custom sizing request from Elementor.

**Comment by amypw** [comment]
@Lukasusanj this solved my issue!! thank you so much. this custom image issue with elementor has been driving me crazy. i guess its an issue with jetpack?

**Comment by cyanit** [comment]
@Lukasusanj tried it, did not work.
Also checked and the site is running on https.

The layout is just screwed:
<img width="944" alt="result" src="https://user-images.githubusercontent.com/26248671/68070220-da75ec00-fd6b-11e9-8d9a-5461b038e3eb.png">

**Comment by Pivtrek** [comment]
> Pro support responded and noted that my home url and site url in the system info file were not the same. I checked both home url and site url in WordPress Settings and they were both exactly the same, like:
> 
> ```
> WordPress Address (URL): http://mysite.com
> Site Address (URL): http://mysite.com
> ```
> 
> I then looked at the system info file and one was https while the other http, like:
> 
> ```
> Site URL:	https://mysite.com	
> Home URL: http://mysite.com
> ```
> 
> This is due to how Cloudflare's Flex SSL works. All urls on the front end are delivered to the browser over https but the website must run under http as no certificate exists on the local web server.
> 
> The problem is not http vs https directly, even though that likely plays a role in how a filter is configured maybe (regex for ex), one can clearly see the file paths and image names are completely different. That is at the core of the problem.
> 
> However I did test this on an http only site without Cloudflare active and I did get it to work on the front end so we may have found the bug. Hopefully it's a quick fix to check for https or Cloudflare plugin to render the correct url. I have tons of my clients that use Cloudflare free SSL and asking them all to go purchase Full SSL certs is not something I want to head up at the moment.
> 
> I responded to the ticket about 17hrs ago, hopefully they find a way to work around this...may even dig into it a bit myself if I get some free time. I'll report back how it goes.

It worked for me, in my case it was by forcing http to https, so the site url and home url were wrong. After changing them to https problem disappeared.

**Comment by plasticpitchfork** [comment]
I have read through this whole thread to see if any of it applies to me. It seems it does not, yet I have the same issue and have had for every site I've tried it on. My site is on a VPS using CPanel Free SSL and the site is configured to https. It is very discouraging to me that Elementor lets these things go for so long. I have not contacted them about it yet, however, as I am not optimistic about their support, based on past experience.

**Comment by plasticpitchfork** [comment]
My server also uses Apache with NGINX caching and a reverse proxy, whatever all that means.

**Comment by rajeev-1974** [comment]
I got the issue resolved by doing following:

1. Go to Wordpress General Settings
2. Add https:// in the beginning of following:
WordPress Address (URL)
Site Address (URL)

**Comment by plasticpitchfork** [comment]
All my sites are configured to https://

**Comment by semy** [comment]
The elementor support is a huge disaster. They only take your money and the rest doesn't matter.

**Comment by semy** [comment]
Could you please do something? This image element still doesn't work. In the editor everything looks fine, on the live page everything sucks!

You can't set width and maximum width.... Neither in pixel nor in %. Fix it finally and do your job and don't push annoying bugs for years. Enough is enough!

@KingYes @kobizz @matipojo @arielk @iNewLegend etc

**Comment by Ife-Ody** [comment]
> Could you please do something? This image element still doesn't work. In the editor everything looks fine, on the live page everything sucks!
> 
> You can't set width and maximum width.... Neither in pixel nor in %. Fix it finally and do your job and don't push annoying bugs for years. Enough is enough!
> 
> @KingYes @kobizz @matipojo @arielk @iNewLegend etc

I've Just tried changing my site URL in the Wordpress admin settings to the HTTPS version eg. http://mysite.com to https://mysite.com and this seems to fix the problem. Check if this solves your problem...

**Comment by walik-walik** [comment]
Same problem here!!!

Have set images for the slider to a size of 300 x 300 in the backend, everything looks wonderful, but not in the frontend!!!! From the looks of it, the problem seems to be a bit older!!!

**Comment by Ife-Ody** [comment]
go to your WordPress settings, and make sure your domain URL is https. this
worked for me

Kind regards,

Ife Odubela.

**Comment by walik-walik** [comment]
Hello, ufff... I thought I had already checked this several times... you were right, that's the problem! Thank you very much

**Comment by Ife-Ody** [comment]
i'm glad you've sorted it out. it took me a long while to realise what the problem was as well even though it is written above haha

**Comment by pluma** [comment]
Just digging this up to add that we had the same problem and changing the URL from http to https in the settings did the trick.

**Comment by BattleGrown** [comment]
For me it was the issue with Jetpack plug-in, disabling image optimization fixed it. So check if you have any issues with SSL or possible tweaks to image hosting. Not sure how elementor can fix this.

**Comment by sammahleruk** [comment]
> For me it was the issue with Jetpack plug-in, disabling image optimization fixed it. So check if you have any issues with SSL or possible tweaks to image hosting. Not sure how elementor can fix this.

That fix helped! thank you very much.

**Comment by baezareich** [comment]
It worked for me to choose another theme in wordpress, hopefully it will help someone...

**Comment by malikatamne** [comment]
> For me it was the issue with Jetpack plug-in, disabling image optimization fixed it. So check if you have any issues with SSL or possible tweaks to image hosting. Not sure how elementor can fix this.

Worked for me too, Thank you!

**Comment by AggieBranderbee** [comment]
Solution: In General settings ensure that both, the Website Address URL and Site Address URL show https: NOT http

That worked for me, showing images resized on both edit and front end.

pheeew!

**Comment by starbuckler** [comment]
This worked for me. Thank you for all the hard-work that went into this solution from users in the past
